### PR TITLE
64 bit migration

### PR DESCRIPTION
**Pull request by MauroToscano**
### Summary

This PR migrates the RISC-V Virtual Machine from 32-bit (RV32IM) to 64-bit (RV64IM) architecture.

#### 64-bit Architecture Migration
- **Registers**: All 32 registers are now 64-bit wide
- **Memory**: Addresses and data operations support 64-bit values
- **Arithmetic**: All base instructions operate on 64-bit values
- **New instructions**: Added RV64-specific instructions:
  - **W-suffix arithmetic**: ADDW, SUBW, ADDIW, MULW, DIVW, DIVUW, REMW, REMUW (operate on lower 32 bits, sign-extend result)
  - **W-suffix shifts**: SLLW, SRLW, SRAW, SLLIW, SRLIW, SRAIW (5-bit shift amount)
  - **64-bit load/store**: LD, SD (doubleword), LWU (load word unsigned/zero-extend)
- **ELF loader**: Updated to parse 64-bit ELF files
- **Shift amounts**: Extended from 5-bit (0-31) to 6-bit (0-63) for base instructions

#### Additional behaviours

**Sign-extension behavior (critical for correctness):**
- `LW` sign-extends loaded word (0x80000000 → 0xFFFFFFFF80000000)
- `LWU` zero-extends loaded word (0x80000000 → 0x0000000080000000)
- `LB`/`LH` sign-extend to 64 bits
- `LBU`/`LHU` zero-extend to 64 bits
- W-suffix results are sign-extended to 64 bits

**Division edge cases (RISC-V spec), now tested**
- `DIV(x, 0)` → -1
- `DIVU(x, 0)` → all-ones
- `REM(x, 0)` → dividend
- `DIV(MIN, -1)` → MIN (overflow returns MIN)
- `REM(MIN, -1)` → 0

### Build System Fixes that were needed during implementation
- Added `std` to `-Z build-std` (was causing infinite loops)
- Changed `max-atomic-width` from 0 to 64 in target spec
- Added `-C passes=lower-atomic` flag (no A extension)
- Added per-program `.cargo/config.toml` with required rustflags
- Fixed CI to use `rustup run nightly cargo` instead of `cargo +nightly`

### Other Changes
- Removed riscv32 references from syscalls (only riscv64 supported)
- Changed `apply_word` panic to proper `Result<_, ExecutionError>`
- Updated CI to use standard `dtolnay/rust-toolchain` instead of custom tarball

### Test Coverage

Added 58 new assembly tests (151 total) covering:
- All W-suffix instructions (ADDW, SUBW, SLLW, MULW, DIVW, etc.)
- 64-bit load/store (LD, SD, LWU)
- Missing branch instructions (BEQ, BLT, BGE, BLTU, BGEU)
- LUI instruction
- 64-bit edge cases (shifts by 32-63, values > 2³², overflow)
- Division by zero and overflow for both 32-bit and 64-bit variants

### Test Results

```
Assembly tests: 150 passed, 1 ignored (unaligned memory access - known limitation)
Rust tests:     24 passed, 1 ignored (ethrex - performance)
```

### CODE REVIEW

File: executor/programs/rust/allocator/.cargo/config.toml [hunk 1]
```diff
@@ -1,5 +1,5 @@
-[target.riscv32im-lambda-vm-elf]
+[target.riscv64im-lambda-vm-elf]
 rustflags = [
-  '--cfg', 'getrandom_backend="custom"',
-  "-C", "passes=lower-atomic" # TODO: check rust version to pass lower-atomic or loweratomic
```
Comment (gianbelinche) [inline executor/programs/rust/allocator/.cargo/config.toml:4]: This flags should still be present, same with all other tests that had them

File: executor/programs/riscv64im-lambda-vm-elf.json [hunk 2]
```diff
@@ -1,18 +1,18 @@
...
-  "data-layout": "e-m:e-p:32:32-i64:64-n32-S128",
+  "data-layout": "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128",
   "eh-frame-header": false,
   "emit-debug-gdb-scripts": false,
   "features": "+m",
   "linker": "rust-lld",
   "linker-flavor": "gnu-lld",
-  "llvm-abiname": "ilp32",
-  "llvm-target": "riscv32",
-  "max-atomic-width": 64,
```
Comment (gianbelinche) [inline executor/programs/riscv64im-lambda-vm-elf.json:13]: From risc0 target:

```
// Some crates (*cough* crossbeam) assume you have 64 bit
            // atomics if the target name is not in a hardcoded list.
            // Since zkvm is singlethreaded and all operations are
            // atomic, I guess we can just say we support 64-bit
            // atomics.
```
I haven't found that issue yet, but we should consider if we want to

File: executor/src/vm/instruction/execution.rs [hunk 3]
```diff
@@ -349,17 +454,19 @@ impl ArithOp {
                     (a as u32).wrapping_rem(b as u32) as i32
                 }
             }
+            // These operations are not valid for W-suffix instructions
+            _ => panic!("Invalid W-suffix operation: {:?}", self),
```
Comment (gianbelinche) [inline executor/src/vm/instruction/execution.rs]: Lets return an error instead of panicking

File: executor/tests/asm.rs [hunk 4]
```diff
@@ -1,12 +1,15 @@
 use executor::{elf::Elf, vm::execution::run_program};
 
-fn run_program_and_check_output(elf_path: &str, expected_output: i32) {
+// NOTE: These tests require 64-bit RISC-V ELF files (RV64IM).
+// The test ELF files need to be recompiled with a 64-bit toolchain.
+// Until then, these tests will fail with "Not a 64-bit ELF" error.
```
Comment (gianbelinche) [inline executor/tests/asm.rs:5]: Lets remove this when the tests pass

File: executor/Makefile [hunk 5]
```diff
@@ -28,31 +28,30 @@ compile-bench: clean-bench $(BENCH_ARTIFACTS)
...
-	clang --target=riscv32 -c $< -o $@
+# Compile and link assembly directly with clang (64-bit)
+$(ASM_ARTIFACTS_DIR)/%.elf: $(ASM_PROGRAMS_DIR)/%.s
+	clang --target=riscv64 -fuse-ld=lld -nostdlib -Wl,-e,main $< -o $@
 
-# Link assembly .o -> .elf
-$(ASM_ARTIFACTS_DIR)/%.elf: $(ASM_ARTIFACTS_DIR)/%.o
-	riscv64-unknown-elf-ld -m elf32lriscv $< -o $@ -e main
+# Custom RV64IM target spec location
+RV64_TARGET_SPEC=$(CURDIR)/programs/riscv64im-lambda-vm-elf.json
```
Comment (gianbelinche) [inline executor/Makefile]: This should be created on our rust fork and used from there (It can be in another pr)

File: executor/Makefile [hunk 6]
```diff
@@ -28,31 +28,30 @@ compile-bench: clean-bench $(BENCH_ARTIFACTS)
...
-		cargo build --release \
-			--target riscv32im-lambda-vm-elf \
-			-Z build-std=core,alloc,std,compiler_builtins,panic_abort \
-			-Z build-std-features=compiler-builtins-mem 
-	cp $(RUST_PROGRAMS_DIR)/$*/target/riscv32im-lambda-vm-elf/release/$* $@
+		cargo +nightly build --release \
+			--target $(RV64_TARGET_SPEC) \
+			-Z build-std=core,alloc,compiler_builtins,panic_abort \
+			-Z build-std-features=compiler-builtins-mem
+	cp $(RUST_PROGRAMS_DIR)/$*/target/riscv64im-lambda-vm-elf/release/$* $@
```
Comment (gianbelinche) [inline executor/Makefile]: Once we have the target in our fork the `+nightly` should be removed

File: executor/Makefile [hunk 7]
```diff
@@ -28,31 +28,30 @@ compile-bench: clean-bench $(BENCH_ARTIFACTS)
...
 $(RUST_ARTIFACTS_DIR)/%.elf: $(RUST_PROGRAMS_DIR)/%/Cargo.toml
 	cd $(RUST_PROGRAMS_DIR)/$* && \
-		cargo build --release \
-			--target riscv32im-lambda-vm-elf \
-			-Z build-std=core,alloc,std,compiler_builtins,panic_abort \
-			-Z build-std-features=compiler-builtins-mem 
-	cp $(RUST_PROGRAMS_DIR)/$*/target/riscv32im-lambda-vm-elf/release/$* $@
+		cargo +nightly build --release \
+			--target $(RV64_TARGET_SPEC) \
+			-Z build-std=core,alloc,compiler_builtins,panic_abort \
```
Comment (gianbelinche) [inline executor/Makefile]: Maybe not having std here would make the tests fail

File: syscalls/src/syscalls.rs [hunk 8]
```diff
@@ -1,11 +1,11 @@
-#[cfg(target_arch = "riscv32")]
+#[cfg(any(target_arch = "riscv32", target_arch = "riscv64"))]
```
Comment (gianbelinche) [inline syscalls/src/syscalls.rs]: We dont really need to keep the `riscv32` since the vm can't emulate it anymore

File: executor/programs/rust/allocator/.cargo/config.toml (hunk 1 above)
Comment (MauroToscano) [reply executor/programs/rust/allocator/.cargo/config.toml:4]: This has been re added

File: executor/Makefile (hunk 6 above)
Comment (MauroToscano) [reply executor/Makefile]: Yes

File: executor/Makefile (hunk 7 above)
Comment (MauroToscano) [reply executor/Makefile]: All test passing one

File: syscalls/src/syscalls.rs (hunk 8 above)
Comment (MauroToscano) [reply syscalls/src/syscalls.rs]: Agreed

File: executor/Makefile [hunk 9]
```diff
@@ -33,54 +38,54 @@ prepare-test-data:
...
-		cargo build --release \
-			--target riscv32im-lambda-vm-elf \
+		CARGO_TARGET_DIR=$(abspath $(SHARED_TARGET_DIR)) \
+		rustup run nightly cargo build --release \
+			--target $(RV64_TARGET_SPEC) \
 			-Z build-std=core,alloc,std,compiler_builtins,panic_abort \
-			-Z build-std-features=compiler-builtins-mem 
-	cp $(RUST_PROGRAMS_DIR)/$*/target/riscv32im-lambda-vm-elf/release/$* $@
+			-Z build-std-features=compiler-builtins-mem
+	cp $(SHARED_TARGET_DIR)/riscv64im-lambda-vm-elf/release/$* $@
```
Comment (gianbelinche) [inline executor/Makefile:63]: Why are we running rustup run?

Comment (MauroToscano) [reply executor/Makefile:63]: Target is missing

**Review by gianbelinche — APPROVED** [review]
